include/zephyr/net/tftp.h

```diff
@@ -53,41 +44,140 @@ struct tftpc {
 #define TFTPC_REMOTE_ERROR       -4
 #define TFTPC_RETRIES_EXHAUSTED  -5
 
-/* @brief This function gets "file" from the remote server.
+/**
+ * @brief TFTP Asynchronous Events notified to the application from the module
+ *        through the callback registered by the application.
+ */
+enum tftp_evt_type {
+	/** DATA event when data is received from remote server.
+	 *
+	 * @note DATA event structure contains payload data and size.
+	 */
+	TFTP_EVT_DATA,
+
+	/** ERROR event when error is received from remote server.
+	 *
+	 * @note ERROR event structure contains error code and message.
+	 */
+	TFTP_EVT_ERROR
+};
+
+/** @brief Parameters for data event. */
+struct tftp_data_param {
+	uint8_t *data_ptr;         /**< Pointer to binary data. */
+	uint32_t len;              /**< Length of binary data. */
+};
+
+/** @brief Parameters for error event. */
+struct tftp_error_param {
+	char *msg;                 /**< Error message. */
+	int code;                  /**< Error code. */
+};
+
+/**
+ * @brief Defines event parameters notified along with asynchronous events
+ *        to the application.
+ */
+union tftp_evt_param {
+	/** Parameters accompanying TFTP_EVT_DATA event. */
+	struct tftp_data_param data;
+
+	/** Parameters accompanying TFTP_EVT_ERROR event. */
+	struct tftp_error_param error;
+};
+
+/** @brief Defines TFTP asynchronous event notified to the application. */
+struct tftp_evt {
+	/** Identifies the event. */
+	enum tftp_evt_type type;
+
+	/** Contains parameters (if any) accompanying the event. */
+	union tftp_evt_param param;
+};
+
+/**
+ * @typedef tftp_callback_t
  *
- * If the file is successfully received its size will be returned in
- * `client->user_buf_size` parameter.
+ * @brief TFTP event notification callback registered by the application.
  *
- * @param server      Control Block that represents the remote server.
- * @param client      Client Buffer Information.
+ * @param[in] evt Event description along with result and associated
+ *                parameters (if any).
+ */
+typedef void (*tftp_callback_t)(const struct tftp_evt *evt);
+
+/**
+ * @brief TFTP client definition to maintain information relevant to the
+ *        client.
+ */
+struct tftpc {
+	/** Pointer to the remote server address structure */
+	struct sockaddr *server;
+
+	/** Receive buffer used for packet in RX path. */
+	uint8_t rx_buf[TFTPC_MAX_BUF_SIZE];
+
+	/** Transmit buffer used for packet in TX path. */
+	uint8_t tx_buf[TFTPC_MAX_BUF_SIZE];
```

Comment:
Allocating 512 bytes for each buffer is quite an overhead. I think that we could just use single buffer. The trick will be to extend `send_data()` API to:
```
static int send_data(int sock, struct tftpc *client, uint32_t block_no,
		     const uint8_t *data, size_t data_size);
```
and then pass user buffer (the one passed to `tftp_put()`) to that function like here:
```
	/* Send out data by chunks */
	do {
		send_size = user_buf_size - tftpc_index;
		if (send_size > TFTP_BLOCK_SIZE) {
			send_size = TFTP_BLOCK_SIZE;
		}
		send_buffer = (uint8_t *)(user_buf + tftpc_index);

		ret = send_data(sock, client, tftpc_block_no, send_buffer, send_size);
```
Then in `send_data()` we can use one buffer for send/recv operations:
```
	/* Send data and poll for ACK response */
	do {
		if (send_count > TFTP_REQ_RETX) {
			LOG_ERR("No more retransmits. Exiting");
			return TFTPC_RETRIES_EXHAUSTED;
		}

		sys_put_be16(DATA_OPCODE, client->buf);
		sys_put_be16(block_no, client->buf + 2);
		memcpy(client->buf + TFTP_HEADER_SIZE, data, data_size);
```
This unfortunately results in more `memcpy()` invocations, however it only happens during retransmission. Even better, `memcpy()` with `send()` could be replaced here with `sendmsg()` that supports gather I/O (from multiple non-contiguous fragments).

The last step would be to update `send_request()` implementation to always create TFTP request, instead of caching it in `client->tx_buf` like that:
```
	do {
		tx_count++;

		LOG_DBG("Sending TFTP request %d file %s", request,
			remote_file);

		/* Create TFTP Request. */
		req_size = make_request(client->buf, request, remote_file, mode);

		/* Send the request to the server */
		ret = sendto(sock, client->buf, req_size, 0, server_addr,
			     ADDRLEN(server_addr));
```

Reply:
Thanks for the suggestion, I get the point. User buffer will be used in PUT in send_data(), and memcy() happens right before transmit or re-transmit to the client buffer (516 bytes). After that using the same client buffer to receive ACK or any possible packet from server.

I'll implement and test on Feb 27th and report back.